File: Foody/app/src/main/java/prateek_gupta/foody/ui/fragments/recipes/bottomSheets/RecipesBottomSheet.java
package prateek_gupta.foody.ui.fragments.recipes.bottomSheets;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveDataReactiveStreams;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.Locale;

import prateek_gupta.foody.R;
import prateek_gupta.foody.databinding.RecipesBottomSheetBinding;
import prateek_gupta.foody.util.Constants;
import prateek_gupta.foody.viewmodels.RecipesViewModel;


public class RecipesBottomSheet extends BottomSheetDialogFragment {
    RecipesViewModel recipesViewModel;

    RecipesBottomSheetBinding binding;

    String mealTypeChip = Constants.DEFAULT_MEAL_TYPE;
    Integer mealTypeChipId = 0;
    String dietTypeChip = Constants.DEFAULT_DIET_TYPE;
    Integer dietTypeChipId = 0;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        recipesViewModel= new ViewModelProvider(requireActivity()).get(RecipesViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding= RecipesBottomSheetBinding.inflate(inflater, container, false);

        LiveDataReactiveStreams.fromPublisher(recipesViewModel.readMealAndDietType).observe(this.getViewLifecycleOwner(),value->{
            mealTypeChip= value.getSelectedMealType();
            dietTypeChip= value.getSelectedDietType();
            updateChip(value.getSelectedMealTypeId(), binding.mealTypeChipGroup);
            updateChip(value.getSelectedDietTypeId(), binding.dietTypeChipGroup);
        });

        binding.mealTypeChipGroup.setOnCheckedStateChangeListener((group, selectedChipId) -> {
            Chip chip= group.findViewById(selectedChipId.get(0));
            mealTypeChip=chip.getText().toString().toLowerCase(Locale.ROOT);
            mealTypeChipId=selectedChipId.get(0);
        });

        binding.dietTypeChipGroup.setOnCheckedStateChangeListener((group, selectedChipId) -> {
            Chip chip= group.findViewById(selectedChipId.get(0));
            dietTypeChip=chip.getText().toString().toLowerCase(Locale.ROOT);
            dietTypeChipId=selectedChipId.get(0);
        });

        binding.applyBtn.setOnClickListener(view -> {
            recipesViewModel.saveMealAndDietType(mealTypeChip,mealTypeChipId,dietTypeChip,dietTypeChipId);
            prateek_gupta.foody.ui.fragments.recipes.bottomSheets.RecipesBottomSheetDirections.ActionRecipesBottomSheetToRecipesFragment action=RecipesBottomSheetDirections.actionRecipesBottomSheetToRecipesFragment();
            action.setBackFromBottomSheet(true);
            Navigation.findNavController(this.requireActivity(),R.id.fragmentContainerView).navigate(action);
        });
        return binding.getRoot();
    }

    void updateChip(Integer chipId, ChipGroup chipGroup) {
        if (chipId != 0) {
            try {
                Chip chip=chipGroup.findViewById(chipId);
                chip.setChecked(true);
            } catch (Exception e ) {
                Log.d("RecipesBottomSheet", e.getMessage());
            }
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding=null;
    }
}

File: Foody/app/src/main/java/prateek_gupta/foody/ui/fragments/recipes/bottomSheets/RecipesBottomSheet.kt
package prateek_gupta.foody.ui.fragments.recipes.bottomSheets

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.asLiveData
import androidx.navigation.fragment.findNavController
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import prateek_gupta.foody.databinding.RecipesBottomSheetBinding
import prateek_gupta.foody.util.Constants.Companion.DEFAULT_DIET_TYPE
import prateek_gupta.foody.util.Constants.Companion.DEFAULT_MEAL_TYPE
import prateek_gupta.foody.viewmodels.RecipesViewModel
import java.util.*

class RecipesBottomSheet : BottomSheetDialogFragment() {
    private lateinit var recipesViewModel: RecipesViewModel

    private var _binding:RecipesBottomSheetBinding?=null
    private val binding get() = _binding!!

    private var mealTypeChip = DEFAULT_MEAL_TYPE
    private var mealTypeChipId = 0
    private var dietTypeChip = DEFAULT_DIET_TYPE
    private var dietTypeChipId = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        recipesViewModel = ViewModelProvider(requireActivity()).get(RecipesViewModel::class.java)
    }
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        _binding= RecipesBottomSheetBinding.inflate(inflater, container, false)
        recipesViewModel.readMealAndDietType.asLiveData().observe(viewLifecycleOwner) { value ->
            mealTypeChip = value.selectedMealType
            dietTypeChip = value.selectedDietType
            updateChip(value.selectedMealTypeId, binding.mealTypeChipGroup)
            updateChip(value.selectedDietTypeId, binding.dietTypeChipGroup)
        }

        binding.mealTypeChipGroup.setOnCheckedStateChangeListener { group, selectedChipId ->
            val chip = group.findViewById<Chip>(selectedChipId[0])
            val selectedMealType = chip.text.toString().lowercase(Locale.ROOT)
            mealTypeChip = selectedMealType
            mealTypeChipId = selectedChipId[0]
        }

        binding.dietTypeChipGroup.setOnCheckedStateChangeListener { group, selectedChipId ->
            val chip = group.findViewById<Chip>(selectedChipId[0])
            val selectedDietType = chip.text.toString().lowercase(Locale.ROOT)
            dietTypeChip = selectedDietType
            dietTypeChipId = selectedChipId[0]
        }

        binding.applyBtn.setOnClickListener {
            recipesViewModel.saveMealAndDietTypeTemp(
                mealTypeChip,
                mealTypeChipId,
                dietTypeChip,
                dietTypeChipId
            )

            val action=RecipesBottomSheetDirections.actionRecipesBottomSheetToRecipesFragment(true)
            findNavController().navigate(action)

        }
        return  binding.root
    }

    private fun updateChip(chipId: Int, chipGroup: ChipGroup) {
        if (chipId != 0) {
            try {
                val targetView = chipGroup.findViewById<Chip>(chipId)
                targetView.isChecked = true
                chipGroup.requestChildFocus(targetView, targetView)
            } catch (e: Exception) {
                Log.d("RecipesBottomSheet", e.message.toString())
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding=null
    }
}